config-service-impl/src/main/java/org/hypertrace/config/service/store/DocumentConfigStore.java

```diff
@@ -76,6 +79,61 @@ public UpsertedConfig writeConfig(
     Optional<ConfigDocument> previousConfigDoc = getLatestVersionConfigDoc(configResourceContext);
     Optional<ContextSpecificConfig> optionalPreviousConfig =
         previousConfigDoc.flatMap(this::convertToContextSpecificConfig);
+
+    Key latestDocKey = new ConfigDocumentKey(configResourceContext);
+    ConfigDocument latestConfigDocument =
+        buildConfigDocument(configResourceContext, latestConfig, userId, previousConfigDoc);
+
+    collection.upsert(latestDocKey, latestConfigDocument);
+    return optionalPreviousConfig
+        .map(previousConfig -> this.buildUpsertResult(latestConfigDocument, previousConfig))
+        .orElseGet(() -> this.buildUpsertResult(latestConfigDocument));
+  }
+
+  private List<UpsertedConfig> writeConfigs(
+      Map<ConfigResourceContext, Value> resourceContextValueMap, String userId) throws IOException {
+    Map<ConfigResourceContext, Optional<ConfigDocument>> previousConfigDocs =
+        getLatestVersionConfigDocs(resourceContextValueMap.keySet());
+    Map<Key, Document> documentsToBeUpserted = new LinkedHashMap<>();
+    for (Entry<ConfigResourceContext, Optional<ConfigDocument>> configResourceContextOptionalEntry :
+        previousConfigDocs.entrySet()) {
+      ConfigResourceContext configResourceContext = configResourceContextOptionalEntry.getKey();
+      Optional<ConfigDocument> previousConfigDoc = configResourceContextOptionalEntry.getValue();
+
+      Key latestDocKey = new ConfigDocumentKey(configResourceContext);
+      ConfigDocument latestConfigDocument =
+          buildConfigDocument(
+              configResourceContext,
+              resourceContextValueMap.get(configResourceContext),
+              userId,
+              previousConfigDoc);
+      documentsToBeUpserted.put(latestDocKey, latestConfigDocument);
+    }
+
+    boolean successfulBulkUpsertDocuments = collection.bulkUpsert(documentsToBeUpserted);
+    if (successfulBulkUpsertDocuments) {
+      return zip(
+              previousConfigDocs.values().stream(),
+              documentsToBeUpserted.values().stream(),
+              (previousConfigDoc, documentToBeUpserted) ->
+                  previousConfigDoc
+                      .flatMap(this::convertToContextSpecificConfig)
+                      .map(
+                          previousConfig ->
+                              this.buildUpsertResult(
+                                  (ConfigDocument) documentToBeUpserted, previousConfig))
```

Comment:
These casts are suspicious - is something mistyped? the zip should not lose the type of values in the stream.

Reply:
`documentsToBeUpserted` isa `Map<Key, Document>` as `bulkUpsert` method expects it to be `Map<Key, Document>`. However, `buildUpsertResult` expects it to be a `ConfigDocument` because of which, having to cast it back.

Reply:
So sounds like a mistype on bulkUpsert - ideally, we'd fix that to accept subtypes (not blocking though).